rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,237 @@
+
+This is where the automation ends. In order for a release (or series of releases) to be deployed to the GOV.UK Production environment, a developer with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) must:
+
+- Manually check the release is "OK" in Integration.
+- Manually deploy the release to Staging.
+- Manually check the release is "OK" in Staging.
+- Manually deploy the release to Production.
+
+Having these tedious, manual steps makes it easy to *not* deploy a change. This is especially true for new joiners, who can only merge PRs and cannot deploy them beyond Integration.
+
+> Full stats and supporting code can be found [here](https://docs.google.com/spreadsheets/d/1Hw1qE-ozqb7n97-fsWkH8Ddgz-7Jyr2G8z8lHXJZPhE/edit#gid=0). We expect these will improve slightly over the coming months, due to the reduction in changes generated by Dependabot [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-126-custom-configuration-for-dependabot.md)].
+
+![Graph of deploy lag (in days)](rfc-128/chart.svg)
+
+A backlog of deployments also makes diagnostics harder:
+
+- ⚠ We can mistakenly debug an issue against the undeployed version of the code.
+- ⚠ We may struggle to isolate a breaking change, if many were deployed together.
+
+### 2. We freely merge PRs with breaking changes
+
+We tend to merge PRs and only later worry if they break the app. This can cause the codebase to become undeployable, and interrupt other developers testing changes on Integration.
+
+Example: https://github.com/alphagov/email-alert-api/commit/8e633fc7d89d7f13bc0e5bdcab14ed3dea968289#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R20
+
+Fix: https://github.com/alphagov/email-alert-api/pull/1346
+
+I couldn't find a good example of myself doing this, but I've definitely done it too. We will never write unbreakable code, yet we have too little caution around merging our changes.
+
+### 3. We do not have a working definition of "OK"
+
+It's easy for each developer to have their own standards, and rely on their own judgement when they check that a change is safe to deploy. Theoretically, we assume a developer will:
+
+- Check the results of [the smoke tests](https://deploy.integration.publishing.service.gov.uk/job/Smokey/).
+- Look for any Icinga alerts related to the app.
+- Check the [deployment dashboard for the app](https://docs.publishing.service.gov.uk/manual/deployment-dashboards.html).
+- Generally make sure the change is tested.
+
+We often defer to automated tests to tell us when something is broken, but it's unclear whether this is really justified: there is no minimum standard for testing of GOV.UK apps.
```

Comment:
I don't think we've adequately described why this a problem. Does this mean we rely on subjective measures of quality or code reliability, hence we don't objectively know if code changes are causing more issues for users in production?

Reply:
Sure, I can explain: aside from the checks listed above, we're supposed to be confident about deploying our changes because our automated tests are passing. But this could be false reassurance if the tests that are passing aren't adequate e.g. if we didn't bother to test something at all (not unheard of). So it's unclear if our deference to PR checks being green is really justified, particularly where we're changing existing code (where we may assume there are also existing tests).

Code coverage is still subjective, in that it's just one way of measuring thoroughness in testing, and the threshold I'm suggesting is also subjective (albeit thoughtfully considered).

So, I'm not sure if there's anything missing here conceptually, but perhaps we can clarify the statement. What do you think?

Reply:
Actually on re-reading this I feel like what we've got there is okay!

Reply:
Thanks, that's OK! I'll resolve this thread now.